src/utils/Logger.cpp

```diff
@@ -9,14 +9,10 @@ namespace Log {
 std::unique_ptr<Logger> global_log;
 
 // Write to stream
-Logger::Logger(logLevel level, std::ostream *os) :
+Logger::Logger(logLevel level, std::shared_ptr<std::ostream> os) :
```

Comment:
Why do you take `os` as a `std::shared_ptr` instead of a reference? This would also remove the need for the workaround in the destructor.
This is more a question out of curiosity than a comment or suggestion.

Reply:
Currently, the workaround is even worse as it is hidden inside the constructor implementation and applied to all streams that are passed - not only the default std::court. 
This change makes the interface safer to use as we do not use raw pointers any more and have to rely on a comment that users have to take care of your passed-over stream's lifetime outside the logger.

Reply:
> A reference is not possible for std::ostream/std:.cout

Maybe I'm missing something obvious, but isn't this how every `operator<<` works?

I just tested this signature:
```cpp
Logger::Logger(logLevel level = Log::Error, std::ostream &os = std::cout) :
```
And changed `_log_stream` to a pointer because the logger is, as far as I understand, not supposed to manage the lifetime of the sink:
```cpp
std::ostream * _log_stream;
```
This compiles and seems to work.

-------------

> This change makes the interface safer to use as we do not use raw pointers any more and have to rely on a comment that users have to take care of your passed-over stream's lifetime outside the logger.

I can see an advantage in the user not having to manage the lifetime of the sink in the `shared_ptr` implementation. At least in the case of ls1 this might indeed be better.